src/wifi/wpa.8
.TH wpa 8
'''
.SH NAME
wpa \- WPA2-PSK client tool (supplicant).
'''
.SH SYNOPSYS
\fBexport\fR PSK\fB=\fR\fI00112233...FF0011...EEFF\fR
.br
wpa \fIdevice\fR \fIfreq\fR \fIbssid\fR "\fIssid\fR" [\fImode\fR]
'''
.SH DESCRIPTION
This tool establishes and maintains an encrypted connection to a given
Wi-Fi access point. Only WPA2-PSK authentication scheme is supported,
with CCMP pairwise cipher and either CCMP or TKIP group ciphers.
.P
The connection \fBwpa\fR deals with is a mac-level (OSI 2) packet transport.
Once the keys have been negotiated, \fIdevice\fR is left with carrier up but
without any higher-level addresses assigned. Other tools like \fBdhcp\fR(8)
should be used to configure IP level connectivity.
'''
.SH USAGE
Generally \fBwpa\fR should be run by a network manager that would handle scans
and AP selection. However, it may be run as a standalone process as well if
the AP is fixed and its parameters are known in advance.
.P
\fBwpa\fR is a long-running process. The connection remains usable for as long
as the process remains alive. Send SIGINT or SIGTERM to \fBwpa\fR to terminate
the connection gracefully. The process may terminate spontaneously if the link
goes down for external reasons, including on AP request.
.P
The following arguments must be passed to \fBwpa\fR:
.IP "\fIdevice\fR" 4
Network device name (\fTwlp1s0\fR).
.IP "\fIfreq\fR" 4
AP base frequency in MHz (\fT5180\fR).
.IP "\fIbssid\fR" 4
MAC address of the AP in hex notation (\fT00:11:22:33:44:55\fR).
.IP "\fIssid\fR" 4
Human-readable AP name ("\fTBlackhole\fR").
.IP "\fImode\fR" 4
Determines cipher suite to use: \fBcc\fR for CCMP pairwise CCMP group or
\fBct\fR CCMP pairwise TKIP group. Defaults to \fBcc\fR if not specified.
.P
The tool expects 32-byte pre-shared key (PSK) in hex notation.
For passphrase-based configuration, the key must be derived from
the passphrase first.
'''
.SH NOTES
Attempts to connect with a wrong key, or using wrong ciphers, are silently
ignored by the AP. Since normal key exchange is very fast, \fBwpa\fR sets a
short timeout and declares a failure if the time runs out.
The lack of error reporting is a feature of the underlying authentication
protocol (EAPOL).
.P
Ciphers and authentication schemes supported by the AP are usually pulled
from the scan results. Doing so is out of scope of this tool. Whoever starts
\fBwpa\fR must be sure the chosen AP does in fact support WPA2-PSK.
.P
Proper WPA2 setups should always use CCMP as both pairwise and group cipher.
However, there are lots of reasonably modern, ostensibly WPA2 APs operating
in WPA1/WPA2 compatibility mode with CCMP pairwise but TKIP group ciphers.
Use \fImode\fR argument to connect with TKIP group cipher if necessary.
.P
This tool does not support WPA2 authentication schemes that aren't WPA2-PSK,
in particular WPA2 Enterprise (including EAP-PSK) and WPA2 FT-PSK (fast
transition with pre-shared key). Nor does it support WPA(1)-PSK. Connections
to APs using either of those schemes must be made with other tools \fIin
place\fR of \fBwpa\fR.
'''
.SH SEE ALSO
\fBwimon\fR(8), \fBwictl\fR(8), \fBwpa_supplicant\fR(8), \fBdhcp\fR(8),
\fBip4cfg\fR(8).


src/wifi/wpa.c
#include <bits/errno.h>
#include <sys/signal.h>
#include <sys/itimer.h>
#include <sys/poll.h>

#include <sigset.h>
#include <string.h>
#include <format.h>
#include <fail.h>
#include <exit.h>
#include <util.h>

#include "wpa.h"

char* ifname;
int ifindex;
int frequency;
uint8_t bssid[6];
char* ssid;

int signalled;
int tkipgroup;

ERRTAG = "wpa";
ERRLIST = {
	REPORT(EINVAL), REPORT(EBUSY), REPORT(ENOENT), REPORT(EPERM),
	REPORT(EFAULT), REPORT(EINTR), REPORT(EINPROGRESS), REPORT(EFAULT),
	REPORT(ENOBUFS), REPORT(ENOLINK), REPORT(ENOTCONN), REPORT(EOPNOTSUPP),
	REPORT(ENETDOWN), REPORT(EALREADY), REPORT(EMSGSIZE), REPORT(EPROTO),
	REPORT(ERANGE), RESTASNUMBERS
};

/* Ref. IEEE 802.11-2012 8.4.2.27 RSNE */

const char ies_ccmp[] = {
	0x30, 0x14, /* ies { type = 48, len = 20 } */
	    0x01, 0x00, /* version 1 */
	    0x00, 0x0F, 0xAC, 0x04, /* CCMP group data chipher */
	    0x01, 0x00, /* pairwise chipher suite count */
	    0x00, 0x0F, 0xAC, 0x04, /* CCMP pairwise chipher */
	    0x01, 0x00, /* authentication and key management */
	    0x00, 0x0F, 0xAC, 0x02, /* PSK and RSNA key mgmt */
	    0x00, 0x00, /* preauth capabilities */
};

const char ies_tkip[] = {
	0x30, 0x14,      /* everything's the same, except for: */
	    0x01, 0x00,
	    0x00, 0x0F, 0xAC, 0x02, /* TKIP group data chipher */
	    0x01, 0x00,
	    0x00, 0x0F, 0xAC, 0x04,
	    0x01, 0x00,
	    0x00, 0x0F, 0xAC, 0x02,
	    0x00, 0x00,
};

const char* ies; /* points to either of above */
int iesize;

static void malarm(int ms)
{
	struct itimerval it = {
		.interval = { 0, 0 },
		.value = { 0, 1000*ms }
	};

	sys_setitimer(ITIMER_REAL, &it, NULL);
}

static void sighandler(int sig)
{
	if(signalled)
		_exit(0xFF);

	signalled = 1;
	malarm(1000);

	const char* msg;

	if(sig == SIGINT || sig == SIGTERM)
		msg = NULL;
	else if(sig == SIGALRM)
		msg = "timeout";
	else
		msg = "signalled";

	quit(msg, NULL, 0);
}

static void setup_signals(void)
{
	int ret = 0;

	struct sigaction sa = {
		.handler = sighandler,
		.flags = SA_RESTORER,
		.restorer = sigreturn,
	};

	sigemptyset(&sa.mask);
	sigaddset(&sa.mask, SIGINT);
	sigaddset(&sa.mask, SIGTERM);
	sigaddset(&sa.mask, SIGHUP);

	ret |= sys_sigaction(SIGINT,  &sa, NULL);
	ret |= sys_sigaction(SIGHUP,  &sa, NULL);
	ret |= sys_sigaction(SIGTERM, &sa, NULL);
	ret |= sys_sigaction(SIGALRM, &sa, NULL);

	if(ret) fail("signal setup failed", NULL, 0);
}

/* This process should be (mildly) privileged to be able to configure
   wdevs, so its envp should not be readable for anyone we should be
   concerned about. However, PSK is still a key, so we wipe it out of
   envp as soon as it gets parsed. */

static void setup_psk(char** envp)
{
	char* arg = getenv(envp, "PSK");
	int len = strlen(arg);
	int req = sizeof(PSK); /* 32 */

	char* p;

	if(!arg)
		fail("no PSK supplied", NULL, 0);
	if(len != 2*req)
		fail("PSK must be 32 bytes long", NULL, 0);
	if(!(p = parsebytes(arg, PSK, req)) || *p)
		fail("invalid PSK", NULL, 0);

	memset(arg, 'x', len); /* PSK=xxxxxx...xxxx */
}

static void mode_ccmp_ccmp(void)
{
	ies = ies_ccmp;
	iesize = sizeof(ies_ccmp);
}

static void mode_ccmp_tkip(void)
{
	ies = ies_tkip;
	iesize = sizeof(ies_tkip);
	tkipgroup = 1;
}

/* envp: PSK=001122...EEFF
   argv: wpa wlp1s0 5180 00:11:22:33:44:55 "Blackhole" [ct]

   Tune wlp1s0 to 5180MHz and connect to station 00:11:22:33:44:55
   named "Blackhole" using supplied PSK. */

static void setup(int argc, char** argv, char** envp)
{
	char* p;
	int i = 1;

	if(i + 4 > argc)
		fail("too few arguments", NULL, 0);

	char* arg_dev = argv[i++];
	char* arg_freq = argv[i++];
	char* arg_bssid = argv[i++];
	char* arg_ssid = argv[i++];
	char* arg_mode = i < argc ? argv[i++] : NULL;

	if(!(p = parsemac(arg_bssid, bssid)) || *p)
		fail("invalid bssid:", arg_bssid, 0);
	if(!(p = parseint(arg_freq, &frequency)) || *p)
		fail("invalid frequency:", arg_freq, 0);

	if(!arg_mode || !strcmp(arg_mode, "cc"))
		mode_ccmp_ccmp();
	else if(!strcmp(arg_mode, "ct"))
		mode_ccmp_tkip();
	else
		fail("invalid mode", arg_mode, 0);

	setup_psk(envp);
	setup_netlink();

	if((ifindex = resolve_ifname(arg_dev)) < 0)
		fail("unknown device", arg_dev, 0);

	ifname = arg_dev;
	ssid = arg_ssid;

	setup_signals();
}

static void poll_netlink_rawsock(void)
{
	int ret;

	struct pollfd fds[2] = {
		{ .fd = netlink, .events = POLLIN },
		{ .fd = rawsock, .events = POLLIN }
	};

	while(1) {
		if((ret = sys_ppoll(fds, 2, NULL, NULL)) < 0)
			quit("ppoll", NULL, ret);

		if(fds[0].revents & ~POLLIN)
			quit("netlink dropped", NULL, 0);
		if(fds[0].revents & POLLIN)
			pull_netlink();

		if(fds[1].revents & ~POLLIN)
			quit("rawsock dropped", NULL, 0);
		if(fds[1].revents & POLLIN)
			if(group_rekey())
				upload_gtk();
	};
}

int main(int argc, char** argv, char** envp)
{
	setup(argc, argv, envp);

	malarm(900);
	authenticate();
	open_rawsock();
	associate();

	malarm(300);
	negotiate_keys();
	upload_ptk();
	upload_gtk();
	cleanup_keys();
	malarm(0);

	poll_netlink_rawsock();

	return 0;
}
